src/Microsoft.SqlTools.ServiceLayer/Copilot/VSCodeChatCompletionStream.cs

```diff
@@ -0,0 +1,192 @@
+        {
+            // Queue the request asynchronously via the channel in ChatMessageQueue
+            request = await CopilotService.Instance.ConversationManager.QueueLLMRequest(
+                this.conversation.ConversationUri, RequestMessageType.ToolCallRequest, messages, tools);
+        }
+
+        public static async Task<VSCodeChatCompletionStream> CreateAsync(
+            CopilotConversation conversation,
+            ChatHistory chat,
+            IList<ChatTool> tools,
+            RequestMessageType requestMessageType)
+        {
+            var collection = new VSCodeChatCompletionStream(conversation, chat, tools, requestMessageType);
+            await collection.InitializeAsync();
+            return collection;
+        }
+
+        private static IList<LanguageModelChatTool> FromChatTools(IList<ChatTool> tools)
+        {
+            return tools.Select(tool => new LanguageModelChatTool
+            {
+                FunctionName = tool.FunctionName,
+                FunctionDescription = tool.FunctionDescription,
+                FunctionParameters = tool.FunctionParameters.SafeToString()
+            }).ToList();
+        }
+
+        private static IList<LanguageModelRequestMessage> FromChatMessages(ChatHistory chatHistory)
+        {
+            var result = new List<LanguageModelRequestMessage>();
+
+            foreach (var message in chatHistory)
+            {
+                if (message.Role == AuthorRole.System || message.Role == AuthorRole.Tool)
+                {
+                    result.Add(new LanguageModelRequestMessage
+                    {
+                        Text = message.Content,
+                        Role = MessageRole.System
+                    });
+                }
+                else if (message.Role == AuthorRole.User)
+                {
+                    if (message.Items is { Count: 1 } &&
+                        message.Items.FirstOrDefault() is TextContent textContent)
+                    {
+                        result.Add(new LanguageModelRequestMessage
+                        {
+                            Text = textContent.Text,
+                            Role = MessageRole.User
+                        });
+                    }
+                    else
+                    {
+                        throw new Exception("Unsupported chat message content type.");
+                    }
+                }
+            }
+
+            // Add explanation message if last message was system
+            if (result.Count > 0 && result[^1].Role == MessageRole.System)
+            {
+                result.Add(new LanguageModelRequestMessage
+                {
+                    Text = "Above messages contain the results of calling a function. " +
+                          "The user cannot see this result, so you should explain it to the user if referencing it in your answer.",
+                    Role = MessageRole.User
```

Comment:
Should this role be `MessageRole.System`? Otherwise this message is showing up in the user messages section.

Reply:
I think there was an issue where the final message in the chat needs to be a user message.  So this was a workaround for that.  This isn't showing up in the users output.

Reply:
Ah I see, I found this line when I was debugging messages. I tested changing this role to `System` and it still worked the same.